### ISSUE TITLE
Idempotency issue with custom role

### ISSUE THREAD

**Issue by LouBea** [IBM-Security/ibmsecurity#69]
Hi Ram,

We have created a role to create many junctions defined in a variable

for example

```
junctions_list: 
  - { set_junction_junction_point: "/aa", set_junction_junction_type: "ssl", set_junction_reverseproxy_id: "{{ fact_isam_proxy_name }}", set_junction_server_hostname: "intg-aa.abc.inet", set_junction_server_port: 465, set_junction_scripting_support: "yes", set_junction_junction_cookie_javascript_block: "trailer", set_junction_remote_http_header: "['iv_groups', 'iv_user']", set_junction_virtual_hostname: "intg-aa.abc.inet:465" }
  - { set_junction_junction_point: "/ab", set_junction_junction_type: "ssl", set_junction_reverseproxy_id: "{{ fact_isam_proxy_name }}", set_junction_server_hostname: "intg-ab.abc.inet", set_junction_server_port: 465, set_junction_scripting_support: "no", set_junction_remote_http_header: "['iv_groups', 'iv_user']", set_junction_virtual_hostname: "intg-ab.abc.inet:465" }
```

the role was created based on set_junction from ibmsecurity but using a "with items" to iterate over many junctions

here is the code (with only the 2 params pertinent to my question)

 
```
#Update the junctions
- name: Set an array of Junctions
  isam:
    appliance: "{{ inventory_hostname }}"
    username:  "{{ username }}"
    password:  "{{ password }}"
    lmi_port:  "{{ lmi_port }}"
    log:       "{{ log_level }}"
    force:     "{{ force }}"
    action: ibmsecurity.isam.web.reverse_proxy.junctions.set
    isamapi:
      ...
      scripting_support:                "{{ item.set_junction_scripting_support | default('no') }}"
      junction_cookie_javascript_block: "{{ item.set_junction_junction_cookie_javascript_block | default([]) }}"
      ...
  with_items: "{{ set_junction_entries }}"
  when: set_junction_entries is defined and item.set_junction_junction_point is defined and item.set_junction_junction_type is defined and item.set_junction_reverseproxy_id is defined and item.set_junction_server_hostname is defined

```

for jct like "aa" (scripting_support = "yes" & junction_cookie_javascript_block = some value ) everything works

for jct like "ab" (scripting_support = "no") I have an idempotence issue

My problem is with junction_cookie_javascript_block param.

when comparing JSONs for the existing junction if scripting_support = "no" the junction_cookie_javascript_block attribute is not returned.

But for the "new" jct because of the "| default([])" it is always in the JSON.

Because of this the 2 JSON are always different & I cannot acheive idempotence for junctions where scripting_support = "no".

In junction.py all other params have

```
if param is None:
    jct_json['param'] = 'default value'
else:
    jct_json['param'] = str(value)
```
so I can use the 'default value' to emulate the behavior of None

but junction_cookie_javascript_block is different:

```
if junction_cookie_javascript_block is not None:
    jct_json['junction_cookie_javascript_block'] = junction_cookie_javascript_block.split(',')
    # Here the list is delimited by space
    if 'junction_cookie_javascript_block' in exist_jct and exist_jct['junction_cookie_javascript_block'] is not None:
        exist_jct['junction_cookie_javascript_block'] = exist_jct['junction_cookie_javascript_block'].split(' ')


```

Would you have a suggestion for how to resolve this problem

Many Thanks
Louis

**Comment by LouBea** [comment]
Hi Ram 
when I use default(None) it is an empty string that is passed to junction's set function, not a None value.

So the issue stays the same

From https://github.com/ansible/ansible/issues/17159 

> When a null variable not inside a datastructure is templated by ansible, it becomes an empty string. The reason is that jinja would otherwise return u'None', so ansible makes use of the "finalize" parameter during instantiation of the jinja environment ...

**Comment by ram-ibm** [comment]
http://docs.ansible.com/ansible/latest/playbooks_filters.html

<cut> In this example, the default None (python null) </cut>

We have been using this. Is it behaving differently because of your Ansible version?

**Comment by LouBea** [comment]
Maybe, we are at ansible 2.4.2.0, in which version have you seen this behave as expected?

also if I change it to `default(omit)` what I get is

`('junction_cookie_javascript_block', ['__omit_place_holder__ea17ac3a3b1f65a5f6c2d98f26b89cebaa46efa1'])`

I just can't seem to find a way for a real None to be passed !

**Comment by LouBea** [comment]
I was able to acheive idempotency by locally tweaking the junction_cookie_javascript_block condition to add the following condition:

```
#We could have a comma delimited set of values - so split them into array
if junction_cookie_javascript_block is not None:
    #to treat an empty string as if it was a None:
    if not( isinstance(junction_cookie_javascript_block, basestring) and len(junction_cookie_javascript_block) == 0 ):
        jct_json['junction_cookie_javascript_block'] = junction_cookie_javascript_block.split(',')
        # Here the list is delimited by space
        if 'junction_cookie_javascript_block' in exist_jct and exist_jct['junction_cookie_javascript_block'] is not None:
            exist_jct['junction_cookie_javascript_block'] = exist_jct['junction_cookie_javascript_block'].split(' ')

```
of course this is just to demonstrate my point & should NOT be a solution for us as we do not want to maintain a customised version of junctions.py

**Comment by ram-ibm** [comment]
I ran a bunch of tests and it does not seem to be related to the version. We are certainly getting an empty string when default(None) jinja2 filter is applied. This may mean I will need to add a `if var is None or var == ''` to all variables being passed.

I am using default(None) in a lot of places - so wonder how this got missed...

**Comment by ram-ibm** [comment]
Turns out we use "null" in default section quite often. That does translate as "None". The issue is with the jinja2 filter. I may need to change logic across the board so that variables that are None OR '' empty string are treated the same. I am betting there will be exceptions - so will need to carefully consider this. Maybe we can just tackle set_junction and then address on a case-by-case basis.

**Comment by LouBea** [comment]
That would be great for us. At this point junction_cookie_javascript_block is the only attribute for which I have foud this empty string/None behavior to cause us idempotency issue.

As you say, they will likely be others but I think addressing them on a case-by-case is very acceptable.

Many thanks for your help!

**Comment by ram-ibm** [comment]
I pushed a change for just junction_cookie_javascript_block - hope you will get a chance to test and close this out please? Thanks.

**Comment by LouBea** [comment]
I tested this modification & its works. Closing.
Thanks for your help!

**Comment by ram-ibm** [comment]
Appreciate your patience - I think this was a valuable use case. Will need to ensure that python code can handle the situation where Ansible passes blank value instead of None. You may also want to reference the code change I made so that future issues like this can leverage a similar code change.

**Comment by LouBea** [comment]
Fixed by bb4400c073063cb0fa0d5c327b49307abd0720a4
